functionary/train_vision/train.py

```diff
@@ -0,0 +1,539 @@
+    else:
+        padding_side = "right"
+
+    # note that must set legacy=True, read more: https://github.com/huggingface/transformers/issues/25176
+    tokenizer = AutoTokenizer.from_pretrained(
+        model_name_or_path,
+        cache_dir=cache_dir,
+        model_max_length=model_max_length,
+        padding_side=padding_side,
+        legacy=True,
+    )
+
+    # Add special tokens
+    tokenizer.pad_token = tokenizer.eos_token
+    added_tokens = prompt_template.get_additional_tokens()
+    special_tokens = {"additional_special_tokens": added_tokens}
+    num_new_tokens = tokenizer.add_special_tokens(special_tokens)
+
+    # add chat_template for tokenizer
+    tokenizer.chat_template = prompt_template.get_chat_template_jinja()
+    print("tokenizer: ", tokenizer)
+
+    # Resize embedding
+    model.resize_token_embeddings(len(tokenizer))
+    if num_new_tokens > 0:
+        input_embeddings = model.get_input_embeddings().weight.data
+        output_embeddings = model.get_output_embeddings().weight.data
+
+        input_embeddings_avg = input_embeddings[:-num_new_tokens].mean(
+            dim=0, keepdim=True
+        )
+        output_embeddings_avg = output_embeddings[:-num_new_tokens].mean(
+            dim=0, keepdim=True
+        )
+
+        input_embeddings[-num_new_tokens:] = input_embeddings_avg
+        output_embeddings[-num_new_tokens:] = output_embeddings_avg
+
+    return tokenizer
+
+
+def extract_unmasked_chunks(labels: List[int], masked_value) -> List[List[int]]:
+    """This function is used to extract unmasked chunks of integer
+    For example, labels = [-100, -100, 1, 2, 3, -100, -100, 4, 5] --> chunks = [[1,2,3], [4,5]]
+    Args:
+        labels (List[int]): list of integer containing token_id and -100
+
+    Returns:
+        List[List[int]]: list of chunk, for example: [[1,2,3], [4,5]]
+    """
+    chunks = []
+    chunk = []
+    for token_id in labels:
+        if token_id != masked_value:
+            chunk.append(token_id)
+        else:
+            if len(chunk) > 0:
+                chunks.append(chunk)
+                chunk = []
+    if len(chunk) > 0:
+        chunks.append(chunk)
+    return chunks
+
+
+def train():
+    argument_parser = transformers.HfArgumentParser(
+        (ModelArguments, DataArguments, TrainingArguments)
+    )
+    model_args, data_args, training_args = argument_parser.parse_args_into_dataclasses()
+    # this is a must
+    training_args.remove_unused_columns = False
+    # this is a must
+    training_args.prompt_template_version = "v3.llava_llama"
+    print(
+        "---------------------training_args.remove_unused_columns: ",
+        training_args.remove_unused_columns,
+    )
+
+    compute_dtype = (
+        torch.float16
+        if training_args.fp16
+        else (torch.bfloat16 if training_args.bf16 else torch.float32)
+    )
+
+    # Copy this from: https://github.com/LLaVA-VL/LLaVA-NeXT/blob/inference/llava/model/builder.py#L169
```

Comment:
Could you use permanent URL instead of latest main branch's Line number

Reply:
what do you mean by permanent URL?  You mean the URL with the commit adding this code?